### PR TITLE
Windows lock timeout

### LINKED ISSUE #59: Sourceforge download links unreliable
A number of packages, I noticed tmux but there are others, have sourceforge links of the form:

```
http://downloads.sourceforge.net/project/tmux/tmux/tmux-1.9/tmux-1.9a.tar.gz
```

These used to work as download links with redirect handling, but now a timestamp appears to be required.  This new form appears to work (on wget from a shell):

```
http://downloads.sourceforge.net/project/tmux/tmux/tmux-1.9/tmux-1.9a.tar.gz?ts=$(date +%s)&use_mirror=autoselect
```

Not sure what the best way to address this is, but a url-handler for sourceforge might be in order.

### PR DESCRIPTION
**Pull request by johnwparent**
Fixes #59 on the Windows Issue Board. 

Solves two issues:

Posix locks can be arbitrarily promoted/demoted by any handle from the same process. This is not true in Windows. Solves the issue by first unlocking the file lock (if necessary) before taking the promotion.

Multiple instances of the `spack.Lock` type take the same lock one after the other are acceptable in Posix locks, but not Windows. Implement a flyweight pattern to serve one lock instance per lock path per process on Windows. This allows all calls to share the same fd handle. The Lock interface should only differ by virtue of calling the factory instead of Lock directly.

### CODE REVIEW

**Comment by johnwparent** [comment]
@BetsyMcPhail PTAL when you get a chance.

File: lib/spack/llnl/util/lock.py
Comment (BetsyMcPhail) [inline lib/spack/llnl/util/lock.py]: Is this line needed?

File: lib/spack/llnl/util/lock.py
Comment (BetsyMcPhail) [inline lib/spack/llnl/util/lock.py]: This is a no-op? Should be `self.current_lock = self.LOCK_EX` ?

**Review by BetsyMcPhail — CHANGES_REQUESTED** [review]

File: lib/spack/llnl/util/lock.py
Comment (johnwparent) [reply lib/spack/llnl/util/lock.py]: Yes it should, thank you.

File: lib/spack/llnl/util/lock.py
Comment (johnwparent) [reply lib/spack/llnl/util/lock.py]: No it is not, missed it during cleanup.

**Review by BetsyMcPhail — APPROVED** [review]
Looks good assuming CI passes

**Comment by johnwparent** [comment]
I have an outstanding branch to verify that locking works on WIndows CI, do we want to merge that in here before merging into the feature branch?

**Comment by BetsyMcPhail** [comment]
> I have an outstanding branch to verify that locking works on WIndows CI, do we want to merge that in here before merging into the feature branch?

I was too quick on the Merge button! If you do a new PR with the locking checks and that fails, we can always revert this change. I'm happy that it's not _more_ broken than the current state.

**Pull request merged**
